### ISSUE TITLE
Reading request body gets blocked by the underlying socket indefinitely.

### ISSUE THREAD

**Issue by yasharpm** [falconry/falcon#1800]
Hi,
I am facing this issue in 2 (out of 2) of my projects using falcon.

Basically what happens is that from time to time reading of request body (either using req.media or req.bounded_stream) gets blocked (not moving to the next line) and since the whole thing is actually running on a single thread, the whole server stops answering any further requests.
By causing a KeyboardInterrupt I can see the stack trace and immediately the buffered requests get handled after wards. Until it happens again.

Here is the stack trace:
```  File "/usr/lib/python3.8/wsgiref/handlers.py", line 137, in run
    self.result = application(self.environ, self.start_response)
  File "/usr/local/lib/python3.8/dist-packages/falcon/api.py", line 269, in __call__
    responder(req, resp, **params)
  File "appheat.py", line 96, in on_post
    for part in media:
  File "/home/appheat/backend/multipart.py", line 441, in __iter__
    stream.pipe_until(delimiter, consume_delimiter=True)
  File "/home/appheat/backend/reader.py", line 323, in pipe_until
    chunk = self._read_until(
  File "/home/appheat/backend/reader.py", line 238, in _read_until
    self._fill_buffer()
  File "/home/appheat/backend/reader.py", line 91, in _fill_buffer
    self._buffer += self._perform_read(read_size)
  File "/home/appheat/backend/reader.py", line 54, in _perform_read
    chunk = self._read_func(size)
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
```

I tried setting a timeout using `httpd.timeout = 1` or setting the socket type to non-blocking using `httpd.socket_type = socket.SOCK_NONBLOCK` and the same thing still happens.

The code is running on Ubuntu 18.0.4 with python version 3.8 and falcon 2. I am connecting to the server directly and there are no web servers or such involved.

_PS. I have taken the multipart code from the unreleased future version of falcon 3 and a bit of patching. I don't believe it is an issue since it is happening at random times and at random positions of the stream._

**Comment by yasharpm** [comment]
Yeah I would happily donate and support the community. It pains me to say that I can neither send or receive money from the unorthodox country I am forced to live in.

**Comment by vytas7** [comment]
Hi @yasharpm !
And thanks for this report.

Would you be able to retest your application with vanilla Falcon 3.0.0a3? There are not that many breaking changes in the WSGI flavour of the framework so, chances are, you might be able to run a smaller app unmodified.
IIRC I had to make a couple of tweaks in the media handling machinery to afford multipart media parsing; you might have missed those in your backport to 2.0.

Another thing, am I right that you are using [`wsgiref.simple_server`](https://docs.python.org/3/library/wsgiref.html#module-wsgiref.simple_server)?

I don't know exactly which version you ported multipart handling from, but could you check if you have something along the lines of
```python
        if not hasattr(stream, 'read_until'):
            if isinstance(stream, request_helpers.BoundedStream):
                stream = BufferedReader(stream.stream.read, content_length)
            else:
                stream = BufferedReader(stream.read, content_length)
```
in your `falcon/media/multipart.py`?
And try simplifying it to:
```python
        if not hasattr(stream, 'read_until'):
            stream = BufferedReader(stream.read, content_length)
```
This should, at a performance cost, always go via `BoundedStream` guard wrapper.
Does that solve the issue?

Furthermore, could you make your test case open somehow? Or at least share it with me, if you can.
I would be happy to debug this, and see if there are any issues in the multipart implementation, itself. We want to fix that ASAP if that turns out to be the case.

Note that the issue could also be simply caused by a malformed request + `wsgiref.simple_server`. If you are sending a wrong `Content-Length` value in your request, to be specific, that is larger than the actual payload size, in my understanding, the `simple_server` can still hang the same way as not using `bounded_stream`.

Last but not the least, feel free to catch me on [`falconry/dev` on Gitter](https://gitter.im/falconry/dev).

cheers
Vytas

**Comment by yasharpm** [comment]
Thank you for making me feel welcome here.

I tried this scenario:
```python
if not hasattr(stream, 'read_until'):
    stream = BufferedReader(stream.read, content_length)
```
And nothing changed.

You are right I was using `wsgiref.simple_server`. So I downloaded Falcon 3.0.0a3 from the repository and switched to ASGI following the instructions in README. So now uvicorn runs the app but I noticed the resource methods are not being called at all. To clear things more, `print` method in the body of the python file works but even deliberately causing fatal exceptions in the resource (`async def on_post()) generates no output. On the client side I receive a connection refused error.

No problem to share the code at all (temporarily though)! [Here is the code](https://github.com/yasharpm/AppHeat_temp). As for the request:
```
POST http://[ip:port]/image?app=com_opentalkz&screen=com_opentalkz_ui_MainActivity&fp=da39a3ee5e6b4b0d3255bfef95601890afd80709
MULTIPART BODY
layoutDescription={"width":900,"height":1550,"elements":[]}
image=[a file]
```

I don't think the problem is from content-length parameter. On the client side it is generated by Android underlying code.

Directly using multi part body implementation of Falcon 3.0.0a3 with WSGI didn't change the outcome either.

**Comment by yasharpm** [comment]
While you're checking my code, I am actually new to Python. My home used to be Java but Python is the best. ;-)

**Comment by vytas7** [comment]
Hi again @yasharpm !
First, a note on `uvicorn`. ASGI is a very different beast, because everything needs to be `async`.
WSGI will still stay as the primary way to develop & deploy Falcon, so it's probably better not to complicate things too much, at least in the context of this issue.
(If you _really_ adventurous, and wanna do it in ASGI instead, multipart form iteration must also by `async`, i.e. `async for part in form: ...`. And you need to `await` I/O, i.e., `await part.get_data()` etc. See more here: https://falcon.readthedocs.io/en/latest/api/multipart.html, check out the "ASGI" tabs.)

Could you install Falcon 3.0.0a3 and Gunicorn as per
```
pip install falcon==3.0.0a3 gunicorn
```
...And then test with this little test case (let's call it `test.py`) instead (feel free to test your app w/ Gunicorn too)?
```python
import wsgiref.simple_server

import falcon


class NewImage:

    def on_post(self, req, resp):
        result = {'content_length': req.content_length, 'parts': []}

        for part in req.media:
            print(f"{part.name} {part.content_type}")
            result['parts'].append((part.name, part.content_type))

            if part.name == 'layoutDescription':
                print(part.text)
            elif part.name == 'image':
                # NOTE: The blank read() is here only for debugging purposes!
                data = part.stream.read()
                print(f"image: {len(data)} byte(s)")

        resp.media = result


app = falcon.App()
app.add_route('/image', NewImage())


if __name__ == '__main__':
    with wsgiref.simple_server.make_server('', 8000, app) as httpd:
        print('Serving on port 8000...')
        httpd.serve_forever()
```
You can run it as
```
gunicorn test:app
```

I was testing it with `curl`:
```
# To get the image :)
# curl --silent https://falcon.readthedocs.io/en/latest/_static/img/logo.svg -o logo.svg

curl -F 'layoutDescription={"message": "Hello"}' -F image=@logo.svg http://localhost:8000/image
```
Feel free to test this example with your client too.

Could you also test your app with curl? I still believe your request may have a malformed form structure, or `Content-Length` header. Someone from the community were reporting issues with the Retrofit library for Android, are you by chance using that?

**Comment by vytas7** [comment]
Another thing. Looking at your feedback wrt going via `BoundedStream`, that sounds like it is not even multipart related.

Could you also try simply dumping your request payload to a file?
Absent better alternatives, you could simple `POST` to the following WSGI app (Falcon 2.0+):
```
import wsgiref.simple_server

import falcon


class Debug:

    def on_post(self, req, resp):
        # NOTE: Adjust the path according to your liking.
        with open('/tmp/debug', 'wb') as output:
            output.write(req.bounded_stream.read())
        resp.media = {'content_length': req.content_length}


app = falcon.API()
app.add_route('/files', Debug())


if __name__ == '__main__':
    with wsgiref.simple_server.make_server('', 8000, app) as httpd:
        print('Serving on port 8000...')
        httpd.serve_forever()

```

**Comment by yasharpm** [comment]
OK I was able to recreate the problem using both test.py and test_dump.py but only through my client app. I couldn't make it happen through Postman.

Focusing on test_dump.py I made sure that the requests I send from my app are identical. About 70% of the time everything works fine. For 30% it blocks with no response being returned hence forward.

Unfortunately when it happens the dump file is empty due read method in `output.write(req.bounded_stream.read())` not returning at all.

Here are the dump samples from Postman and my app so you can compare:
[Dump from PostMan](https://github.com/yasharpm/AppHeat_temp/raw/master/postman)
[Dump from App](https://github.com/yasharpm/AppHeat_temp/raw/master/app)

I think it is a good opportunity for me to point out that this event also happens with my other Falcon project. But it happens about once a month in which I'll have to restart. I guess it is somehow dependent on the request body size.

**Comment by vytas7** [comment]
Thanks for this!
As discussed on Gitter, we also need to test this using Gunicorn (or any other production-level WSGI application server).

Although I don't think Falcon as a framework can do much here, it would be interesting to get to the bottom of this issue.
If it not something caused by your client, it would be good to document this for others that may stumble upon similar problems in `wsgiref.simple_server` (or whatever the real cause is).

**Comment by yasharpm** [comment]
After using WSGI the same thing happens except this time reading from the socket actually times out and so the new requests get way to be processed. I don't know why data stream should suddenly stop (maybe it's internet filtering related), but now that the code can get over it, it is actually no problem any more.

Thank you very much for bearing with me and helping to solve this.

**Comment by vytas7** [comment]
Thank your for this whole thread too.
This is could loosely be classified as a duplicate of #1483 (which I've just reopened).
